content/articles/hashing-in-data-structure/index.md

```diff
@@ -0,0 +1,299 @@
+- **Fast**: The generation of hash should be very fast, and should not produce any considerable overhead.
+
+#### Collision Hashing Techniques:
+
+1. **Open Hashing (Separate Chaining)**: In this using Linked List, new nodes are added to the list, the key act as the head pointer, pointing to several nodes having different values. If any collision two elements is having same hash value we store both of them in a same linked list.
+
+Let's consider an example of a simple hash function	
+   
+   `h(key) = key%table size`
+
+   In a hash table with the size 7
+   
+   h(27) = 27%7 = 6
+   
+   h(130) = 130%7 = 4
+
+   ![Example](./hashMap_1.PNG)
+
+   If we insert a new element (18, "Saleema" ), that would also go to the 4th index.
+   
+   h(18) = 18%7 = 4
+
+   ![Example](./hashMap_2.PNG);
+
+   For separate chaining, the worst-case scenario is when all of the keys will get the same hash value and will be inserted in the same linked list. We can avoid this by using a good hash function.
+
+2. **Closed Hashing (Open Addressing)**: In this, we find the "next" vacant bucket in Hash Table and store the value in that bucket.
+
+   1. **Linear Probing**: We linearly go to every next bucket and see if it is vacant or not.
+  
+      `rehash(key) = (n+1)%tablesize`
+
+   2. **Quadratic Probing**: We go to the 1st, 4th, 9th ... bucket and check if they are vacant or not.
+   
+      `rehash(key) = (n+ k<sup>2</sup> ) % tablesize
+
+3. **Double Hashing**: Here we subject the generated key from the hash function to a second hash function.
+
+   ` h2(key) != 0 and h2 != h1`
+
+**Load Factor**: This is a measurement of how full a hash table may become before its capacity is increased. 
+
+The hash table's load factor, T, is defined as:
+
+- N = number of elements in T - Current Size
+- M = size of T - Table Size
+- e = N/M - Load factor
+
+Generally, if the load factor is greater than 0.5, we increase the size of the bucket array and rehash all the key-value pairs again.
+
+#### How Hashing Gets O(1) Complexity?
+
+Given the above examples, one would wonder how hashing may be O(1) if several items map to the same place...  
+
+The solution to this problem is straightforward. We use the load factor to ensure that each block, for example, (linked list in a separate chaining strategy), stores the maximum amount of elements fewer than the load factor on average. Also, in practice, this load factor is constant ( generally 10 or 20 ). As a result, searching in 10 or 20 elements become constant.
+
+If the average number of items in a block exceeds the load factor, the elements are rehashed with a larger hash table size.
+
+**Rehashing**
+
+When the load factor gets "too high" (specified by the threshold value), the process of raising the size of the hash table begins, expecting that collisions would become more common.
+
+- Typically, double the size of the table (but still prime).
+- All existing items must be reinserted into the hash table.
+
```

Comment:
This section is not clear. Are they the solutions to the load factor getting "too high"? Please make it clear